### PR TITLE
#6 Add Digital Ocean Domain Records

### PR DESCRIPTION
**Pull request by rezen**
##### SUMMARY
Add a module to manage DO domain records, copied PR https://github.com/ansible/ansible/pull/31765/files, changing only lines noted from comments

Addresses https://github.com/ansible-collections/community.digitalocean/issues/6

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
`digital_ocean_domain_record`

### CODE REVIEW

**Comment by rezen** [comment]
Should I rename the module domain_records?

**Comment by mamercad** [comment]
> Should I rename the module domain_records?

Probably a good idea in the spirit of https://github.com/ansible-collections/community.digitalocean/pull/27.

**Comment by jamesharr** [comment]
@mamercad is there anything needed to polish this module up for merge?

**Comment by mamercad** [comment]
> @mamercad is there anything needed to polish this module up for merge?

It looks good to me, I'd like @bmildren to take a peek though please!

**Comment by clucia** [comment]
I'm using this, and it works with the exception that it's not idempotent. To work around that, I test for the existence of the domain records in the yaml and delete them and re-create them every time.

**Comment by rezen** [comment]
Sounds like the `create_or_update_record` is not telling the truth about itself?

**Comment by clucia** [comment]
This domain record is in my list:
139034089    A       test    10.0.0.99         0           0       144     0k

Here's what's in my playbook to delete it:
  - name: domain
    community.digitalocean.domain_record:
      name: test
      domain: googular.com
      type: A
      state: absent

It's not deleting the record. 

on the other hand, this playbook seems to work:

  - name: domain
    community.digitalocean.domain_record:
      name: test
      domain: googular.com
      state: present
      force_update: yes
      type: A
      data: 10.0.0.99
      ttl: 144

I was previously not using the force_update option.

**Comment by rezen** [comment]
@clucia if you tweak your configurations a bit `absent` will work as expected. You have to include the full details of the record you want absent. 

Since you can have multiple `A` records for a domain, how do differentiate between I want only this record to exist vs I need to add a new `A` record? Making you use `force_update` seems to be a "protective control" to prevent removing/changing existing records when you want a new record.

```
name: domain
community.digitalocean.domain_record:
     name: test
     domain: googular.com
-    state: present
+    state: absent
     force_update: yes
     type: A
     data: 10.0.0.99
     ttl: 144
```

**Comment by mamercad** [comment]
Working through a few things here: https://github.com/rezen/community.digitalocean/pull/1

**Comment by mamercad** [comment]
Idempotence seems fine: https://gist.github.com/mamercad/10773e19f28e88693a331ad5f486a7a7

**Comment by mamercad** [comment]
@rezen Want to add a quick [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment)?

File: plugins/modules/digital_ocean_domain_record.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+
+        return payload
+
+    def delete_record(self):
+
+        # if record_id is given, try to find the record based on the id
+        if self.record_id:
+            record = self.__find_record_by_id(self.record_id)
+            record_id = self.record_id
+        # if no record_id is given, try to find an exact match
```
Comment (sgpinkus) [inline plugins/modules/digital_ocean_domain_record.py:388]: I'm using this already. Thanks. One thing is `state: present` matches records differently to `state: absent`. This means if you install a domain record then simply flip `state: present` to `state: absent` the module will not find the record you just created. `state: present` and `state: absent` need to use the same logic to locate an existing record.

Comment (sgpinkus) [reply plugins/modules/digital_ocean_domain_record.py:388]: Suggested change: https://github.com/rezen/community.digitalocean/pull/3/files

**Comment by rezen** [comment]
@mamercad added the changelog fragment - not sure if I did that right?

**Comment by mamercad** [comment]
> @mamercad added the changelog fragment - not sure if I did that right?

Based on the docs (linked above):

Major changes are "changes to Ansible itself. Generally does not include module or plugin changes. Displayed in both the changelogs and the Porting Guides"

Minor changes are "changes to Ansible, modules, or plugins. This includes new features, new parameters added to modules, or behavior changes to existing parameters".

It seems like minor changes would be more appropriate. In the docs, they also go on to say "most changelog entries will be bugfixes or minor_changes".

So, I think we want `minor_changes` for this.

Lastly, I think it needs to look like this (including the module name):

```yaml
minor_changes:
  - digital_ocean_domain_record - Added plugin module for managing domain records
```

Thoughts on this @mpontillo ?

**Review by mamercad — APPROVED** [review]

**Comment by mamercad** [comment]
I'll add the fragment in a release branch, no biggie, thanks so much for this module!

**Comment by rezen** [comment]
Woohoo! 🎉

**Pull request merged**